sound/soc/sof/pm.c

```diff
@@ -16,8 +16,146 @@
 #include <asm/pgtable.h>
 #include <sound/core.h>
 #include <sound/soc.h>
+#include <sound/sof.h>
+#include "ops.h"
 #include "sof-priv.h"
 
+static int sof_restore_pipeline(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_widget *swidget = NULL;
+	struct snd_sof_route *sroute = NULL;
+	struct sof_ipc_comp_host *host;
+	struct sof_ipc_buffer *buffer;
+	struct sof_ipc_comp_volume *vol;
+	struct sof_ipc_comp_mixer *mixer;
+	struct sof_ipc_pipe_new *pipeline;
+	struct sof_ipc_comp_src *src;
+	struct sof_ipc_comp_tone *tone;
+	struct snd_sof_dai *dai;
+	struct sof_ipc_comp_dai *comp_dai;
+	int ret = 0;
+
+	/* restore pipeline components */
+	list_for_each_entry_reverse(swidget, &sdev->widget_list, list) {
+		struct sof_ipc_comp_reply r;
+
+		switch (swidget->id) {
+		case snd_soc_dapm_dai_in:
+		case snd_soc_dapm_dai_out:
+			dai = (struct snd_sof_dai *)swidget->private;
```

Comment:
Just wondering if we need all these case statements given all the IPC data uses a standard header (which includes cmd and size). i.e. we cas the private data to the header type to get size and cmd.

Reply:
@lgirdwood I started it that way but I kept running into errors with it. The only way I can send the ipc without errors is by having all these case statements.

Reply:
ok, interesting, what errors ? The only thing the case statements do is cast types

Reply:
@lgirdwood. I've fixed it now. The error I was seeing was because of not casting the comp objects to (void *) while storing them.